lib/ACL/ACLManager.php

```diff
@@ -54,40 +54,43 @@ private function getRootStorageId() {
 		return $this->rootStorageId;
 	}
 
-	private function pathsAreCached(array $paths): bool {
-		foreach ($paths as $path) {
-			if (!$this->ruleCache->hasKey($path)) {
-				return false;
-			}
-		}
-		return true;
-	}
-
 	/**
 	 * @param int $folderId
 	 * @param array $paths
 	 * @return (Rule[])[]
 	 */
 	private function getRules(array $paths): array {
-		if ($this->pathsAreCached($paths)) {
-			$rules = array_combine($paths, array_map(function (string $path) {
-				return $this->ruleCache->get($path);
-			}, $paths));
-		} else {
-			$rules = $this->ruleManager->getRulesForFilesByPath($this->user, $this->getRootStorageId(), $paths);
-			foreach ($rules as $path => $rulesForPath) {
+		// beware: adding new rules to the cache besides the cap
+		// might discard former cached entries, so we can't assume they'll stay
+		// cached, so we read everything out initially to be able to return it
+		$rules = array_combine($paths, array_map(function (string $path) {
+			return $this->ruleCache->get($path);
+		}, $paths));
+
+		$nonCachedPaths = array_filter($paths, function (string $path) use ($rules) {
+			return !isset($rules[$path]);
+		});
+
+		if (!empty($nonCachedPaths)) {
+			$newRules = $this->ruleManager->getRulesForFilesByPath($this->user, $this->getRootStorageId(), $nonCachedPaths);
+			foreach ($newRules as $path => $rulesForPath) {
 				$this->ruleCache->set($path, $rulesForPath);
+				$rules[$path] = $rulesForPath;
 			}
+		}
 
-			if (count($paths) > 2) {
-				// also cache the direct sibling since it's likely that we'll be needing those later
-				$directParent = $paths[1];
+		if (count($paths) > 2) {
+			// also cache the direct sibling since it's likely that we'll be needing those later
+			$directParent = $paths[1];
+			if (!isset($rules[$directParent])) {
```

Comment:
I still do not understand, if `ruleCache->get($paths[1])` is null, `$paths[1]` is put in `$nonCachedPaths`, and `ruleManager->getRulesForFilesByPath` is called on it on line 75, and the result is stored in `$rules`.
Is it still possible for `$rules[$paths[1]]` to not be set after that?

Reply:
I've just investigated that with a debugger, and after calling `getRulesFoFilesByPath` I saw that all the keys will be populated. If no rules, there will be an empty array: https://github.com/nextcloud/groupfolders/blob/bug/1693/fix-acl-perfs/lib/ACL/RuleManager.php#L116

but I get your point, I also wonder why we'd need that second block at all. I'll recheck with the original code

Reply:
alright, got my mistake. the block isn't loading a rule for `$directPath` but it's doing so for the siblings.
the call is on `getRulesForFilesByParent` not `getRulesForFilesByPath`.

since I don't know the sibling names I can't do a cache check, so I'll simply let it through and hope the perf issue doesn't appear here after testing